### PR TITLE
invoice: Support providing preimage when making invoice.

### PR DESCRIPTION
**Pull request by ZmnSCPxj**
Fixes: #1402

### CODE REVIEW

File: doc/lightning-invoice.7.txt [hunk 1]
```diff
@@ -36,6 +36,20 @@ The 'fallbacks' array is one or more fallback addresses to include in
...
+The 'preimage' is a 64-digit hex string to be used as payment preimage
+for the created invoice.
+By default, if unspecified, lightningd will generate a secure
+pseudorandom preimage seeded from an appropriate entropy source on
+your system.
+*IMPORTANT*: if you specify the 'preimage', you are responsible, to
+ensure appropriate care for generating using a secure pseudorandom
+generator seeded with sufficient entropy, and keeping the preimage
+secret.
+This parameter is intended for use with ZKCP protocols, where the
```
Comment (cdecker) [inline doc/lightning-invoice.7.txt]: This is maybe a bit over specific, mentioning that it is an advanced feature and should not be used unless explicitly needed should suffice here.

File: lightningd/invoice.c [hunk 2]
```diff
@@ -281,8 +283,18 @@ static void json_invoice(struct command *cmd,
 	struct preimage r;
 	struct sha256 rhash;
 
-	/* Generate random secret preimage and hash. */
-	randombytes_buf(r.r, sizeof(r.r));
+	if (preimagetok) {
+		/* Get secret preimage from user. */
+		if (!hex_decode(buffer + preimagetok->start,
```
Comment (cdecker) [inline lightningd/invoice.c]: If I'm not mistaken this does not raise an error when we provide less than 64 digits.

Comment (ZmnSCPxj) [reply lightningd/invoice.c]: It does: https://github.com/ElementsProject/lightning/blob/9c3691340fa430ed1daf08dc40f76bd7369f66cc/ccan/ccan/str/hex/hex.c#L24-L40

Specifically: https://github.com/ElementsProject/lightning/blob/9c3691340fa430ed1daf08dc40f76bd7369f66cc/ccan/ccan/str/hex/hex.c#L39

If `bufsize == 0` is untrue, or `slen` is odd, then the condition at `return` will fail. At least, my understanding of the code.

File: doc/lightning-invoice.7.txt (hunk 1 above)
Comment (ZmnSCPxj) [reply doc/lightning-invoice.7.txt]: Okay, will modify.

**Comment by cdecker** [comment]
ACK fdd30b1

**Comment by mcudev** [comment]
Just a note that the unique constraint causes node crash.

Sequence of commands:
```
$ ./lightning/cli/lightning-cli --lightning-dir=/tmp/l1 -k invoice msatoshi=1000000000 label=i1 description=i1 preimage=0000000000000000000000000000000000000000000000000000000000000000
{
  "payment_hash": "66687aadf862bd776c8fc18b8e9f8e20089714856ee233b3902a591d0d5f2925", 
  "expiry_time": 1524608889, 
  "expires_at": 1524608889, 
  "bolt11": "lnbcrt10m1pddlgtfpp5ve584t0cv27hwmy0cx9ca8uwyqyfw9y9dm3r8vus9fv36r2l9yjsdqydycscqpxnqa306naqzd5mj3x8v6yswwzavmexxmz896weuj8mvm8cd7t8n65s3x8thqxv0e9c7uf3jdjcshu74sq733lw5cz8cjy5fxuvf6xdngqdvaxad"
}
$ ./lightning/cli/lightning-cli --lightning-dir=/tmp/l1 -k invoice msatoshi=1000000000 label=i2 description=i2 preimage=0000000000000000000000000000000000000000000000000000000000000000
lightning-cli: reading response: Success
```

log entries:
```
2018-04-24T21:28:29.828Z lightningd(19347): invoices_create: UNIQUE constraint failed: invoices.payment_hash
2018-04-24T21:28:29.841Z lightningd(19347): FATAL SIGNAL 6
2018-04-24T21:28:29.841Z lightningd(19347): backtrace: common/daemon.c:42 (crashdump) 0x432c43
2018-04-24T21:28:29.841Z lightningd(19347): backtrace: (null):0 ((null)) 0x7f0a3b2e04af
2018-04-24T21:28:29.841Z lightningd(19347): backtrace: (null):0 ((null)) 0x7f0a3b2e0428
2018-04-24T21:28:29.841Z lightningd(19347): backtrace: (null):0 ((null)) 0x7f0a3b2e2029
2018-04-24T21:28:29.841Z lightningd(19347): backtrace: lightningd/log.c:553 (fatal) 0x41606d
2018-04-24T21:28:29.841Z lightningd(19347): backtrace: wallet/db.c:332 (db_exec_prepared_) 0x45668b
2018-04-24T21:28:29.841Z lightningd(19347): backtrace: wallet/invoices.c:309 (invoices_create) 0x45887a
2018-04-24T21:28:29.841Z lightningd(19347): backtrace: wallet/wallet.c:1437 (wallet_invoice_create) 0x45de2a
2018-04-24T21:28:29.841Z lightningd(19347): backtrace: lightningd/invoice.c:316 (json_invoice) 0x40efd3
2018-04-24T21:28:29.841Z lightningd(19347): backtrace: lightningd/jsonrpc.c:475 (parse_request) 0x4121cb
2018-04-24T21:28:29.841Z lightningd(19347): backtrace: lightningd/jsonrpc.c:645 (read_json) 0x412a95
2018-04-24T21:28:29.841Z lightningd(19347): backtrace: ccan/ccan/io/io.c:59 (next_plan) 0x468cb4
2018-04-24T21:28:29.841Z lightningd(19347): backtrace: ccan/ccan/io/io.c:387 (do_plan) 0x46978b
2018-04-24T21:28:29.841Z lightningd(19347): backtrace: ccan/ccan/io/io.c:397 (io_ready) 0x4697c9
2018-04-24T21:28:29.841Z lightningd(19347): backtrace: ccan/ccan/io/poll.c:310 (io_loop) 0x46b335
2018-04-24T21:28:29.841Z lightningd(19347): backtrace: lightningd/lightningd.c:399 (main) 0x41454a
2018-04-24T21:28:29.841Z lightningd(19347): backtrace: (null):0 ((null)) 0x7f0a3b2cb82f
2018-04-24T21:28:29.842Z lightningd(19347): backtrace: (null):0 ((null)) 0x403488
2018-04-24T21:28:29.842Z lightningd(19347): backtrace: (null):0 ((null)) 0xffffffffffffffff
```

**Pull request merged**